### PR TITLE
rustdoc: Correct anchor for links to associated trait items

### LINKED ISSUE #28478: rustdoc generates bad fragments in links to trait methods
For example, https://doc.rust-lang.org/nightly/std/net/enum.SocketAddr.html links to "https://doc.rust-lang.org/nightly/std/fmt/trait.Display.html#method.fmt", but the correct link is "https://doc.rust-lang.org/nightly/std/fmt/trait.Display.html#tymethod.fmt".

### PR DESCRIPTION
**Pull request by mitaa**
fixes #28478

r? @alexcrichton

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -2085,7 +2085,13 @@ fn render_assoc_item(w: &mut fmt::Formatter, meth: &clean::Item,
         let href = match link {
             AssocItemLink::Anchor => anchor,
             AssocItemLink::GotoSource(did) => {
-                href(did).map(|p| format!("{}{}", p.0, anchor)).unwrap_or(anchor)
+                // We're creating a link from an impl-item to the corresponding
+                // trait-item and need to map the anchored type accordingly.
```
Comment (mitaa) [inline src/librustdoc/html/render.rs]: Ah, I see. This is wrong for default methods ("provided methods")

```rust
pub trait Bar {
    // #method.bar
    fn bar(&self) {}
    // #tymethod.baz
    fn baz(&self);
}

pub struct Foo;

impl Bar for Foo {
    fn baz(&self) {}
}
```

**Comment by alexcrichton** [comment]
Hm so right now we have a [linkchecker script](https://github.com/rust-lang/rust/blob/master/src/tools/linkchecker/main.rs) which ensures that all links point to valid locations. Perhaps we could extend that with checking anchors as well? It may be relatively easy enough to just grep for `id="foo"` in a page given an anchor that we see.

**Comment by mitaa** [comment]
Good idea, I've mostly got the linkchecker working, now I just need to somehow deal with redirect pages.
(going to push this onto this branch sometime today)

**Comment by alexcrichton** [comment]
Nice! Feel free to ping me when that happens and I'll likely r+ :)

**Comment by mitaa** [comment]
(updated)

I've had to whitelist some cases in the same vein as #32130, and `collections/string/String` does not have the `methods from Deref<Target=str>` section for some reason breaking links to `str` methods on the same page (not sure what's up with that).

File: src/tools/linkchecker/main.rs [hunk 2]
```diff
@@ -45,74 +47,262 @@ fn main() {
...
 
-    let pretty_file = file.strip_prefix(root).unwrap_or(file);
-
     // Unfortunately we're not 100% full of valid links today to we need a few
     // whitelists to get this past `make check` today.
     // FIXME(#32129)
-    if file.ends_with("std/string/struct.String.html") {
-        return
+    if file.ends_with("std/string/struct.String.html") ||
+       file.ends_with("collections/string/struct.String.html") {
```
Comment (alexcrichton) [inline src/tools/linkchecker/main.rs]: For the addition of these new ignores, can you be sure that issues are filed and these are tagged with a FIXME?

**Comment by alexcrichton** [comment]
@bors: r+ 41916d860179f4715fd0f588c1b2342a223a4186

Oh man this is exciting, thanks @mitaa!

**Comment by mitaa** [comment]
(updated, should work now)

**Pull request merged**